File: interaction_state.mjs
import crypto from 'crypto';
import fs from 'fs';
import path from 'path';

import {
  VALID_LANGUAGES,
  reduceLanguageState,
  sanitizeLanguageState,
} from './language_adaptation.mjs';

const VALID_KINDS = new Set(['call', 'content']);

function fingerprint(namespace, value) {
  return crypto.createHash('sha256').update(`${namespace}\0${String(value)}`, 'utf8').digest('hex');
}

export class PersistentInteractionState {
  constructor({
    filePath,
    defaultLanguage = 'es',
    maxRecentEvents = 256,
    now = () => Date.now(),
    logger = console,
  }) {
    if (!filePath) throw new TypeError('filePath is required');
    if (!VALID_LANGUAGES.has(defaultLanguage)) {
      throw new TypeError('defaultLanguage must be es, en, or fr');
    }
    if (!Number.isInteger(maxRecentEvents) || maxRecentEvents < 1) {
      throw new TypeError('maxRecentEvents must be a positive integer');
    }

    this.filePath = filePath;
    this.defaultLanguage = defaultLanguage;
    this.maxRecentEvents = maxRecentEvents;
    this.now = now;
    this.logger = logger;
    this.contacts = {};
    this.aliases = {};
    this.load();
  }

  load() {
    if (!fs.existsSync(this.filePath)) return;
    try {
      const parsed = JSON.parse(fs.readFileSync(this.filePath, 'utf8'));
      if (!parsed.contacts || typeof parsed.contacts !== 'object' || Array.isArray(parsed.contacts)) {
        throw new TypeError('contacts is not an object');
      }

      for (const [contactKey, raw] of Object.entries(parsed.contacts)) {
        if (!raw || typeof raw !== 'object' || ![0, 1, 2].includes(raw.phase)) continue;
        const resetPending = raw.reset_pending === true && raw.phase === 0;
        this.contacts[contactKey] = {
          phase: raw.phase,
          ...sanitizeLanguageState(raw),
          recent_events: !resetPending && Array.isArray(raw.recent_events)
            ? raw.recent_events.filter(item => typeof item === 'string').slice(-this.maxRecentEvents)
            : [],
          updated_at: Number.isFinite(raw.updated_at) ? raw.updated_at : 0,
          ...(resetPending ? { reset_pending: true } : {}),
        };
      }
      if (parsed.aliases && typeof parsed.aliases === 'object' && !Array.isArray(parsed.aliases)) {
        for (const [aliasKey, contactKey] of Object.entries(parsed.aliases)) {
          if (typeof contactKey === 'string' && this.contacts[contactKey]) {
            this.aliases[aliasKey] = contactKey;
          }
        }
      }
    } catch {
      this.logger.error?.('[STATE] Interaction state could not be loaded; starting empty');
      this.contacts = {};
      this.aliases = {};
    }
  }

  save() {
    try {
      fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
      const temporaryPath = `${this.filePath}.tmp`;
      fs.writeFileSync(
        temporaryPath,
        JSON.stringify({ version: 2, contacts: this.contacts, aliases: this.aliases }),
        'utf8',
      );
      fs.renameSync(temporaryPath, this.filePath);
      return true;
    } catch {
      this.logger.error?.('[STATE] Interaction state could not be persisted');
      return false;
    }
  }

  register({
    contactId,
    contactAliases = [],
    eventId,
    kind,
    detectedLanguage = null,
    languageEvidence = null,
    provisionalLanguage = null,
  }) {
    if (contactId === undefined || contactId === null || contactId === '') {
      throw new TypeError('contactId is required');
    }
    if (eventId === undefined || eventId === null || eventId === '') {
      throw new TypeError('eventId is required');
    }
    if (!VALID_KINDS.has(kind)) throw new TypeError('kind must be call or content');
    if (!VALID_LANGUAGES.has(provisionalLanguage)) provisionalLanguage = null;

    const identityValues = [contactId, ...(Array.isArray(contactAliases) ? contactAliases : [])]
      .filter(value => value !== undefined && value !== null && value !== '');
    const identityKeys = [...new Set(identityValues.map(value => fingerprint('contact', value)))];
    const primaryKey = identityKeys[0];
    const eventKey = fingerprint('event', eventId);
    // Keep both a direct contact and its alias target. During the short PN/LID
    // convergence window both records may legitimately exist.
    const resolvedKeys = [...new Set(identityKeys.flatMap(key => [key, this.aliases[key]]))]
      .filter(key => typeof key === 'string');
    const existingKeys = resolvedKeys.filter(key => this.contacts[key]);
    const pendingResetKey = existingKeys.find(
      key => this.contacts[key].reset_pending === true && this.contacts[key].phase === 0,
    );
    const canonicalKey = pendingResetKey || existingKeys[0] || primaryKey;
    const states = existingKeys.map(key => this.contacts[key]);
    const pendingResetState = pendingResetKey ? this.contacts[pendingResetKey] : null;
    // A duplicate is observational: it must not merge identities, consume a
    // reset marker, advance a candidate, or rewrite aliases.
    const duplicateStates = pendingResetState ? [pendingResetState] : states;
    const duplicateState = duplicateStates.find(candidate => (
      Array.isArray(candidate.recent_events) && candidate.recent_events.includes(eventKey)
    ));
    if (duplicateState) {
      return {
        duplicate: true,
        phase: duplicateState.phase,
        responseKey: null,
        language: duplicateState.language || this.defaultLanguage,
        languageSource: duplicateState.language_source || null,
        languageCandidate: duplicateState.language_candidate || null,
        languageCandidateStreak: duplicateState.language_candidate_streak || 0,
        contactKey: canonicalKey,
        persisted: true,
      };
    }
    const state = pendingResetState || states[0] || {
      phase: 0,
      ...sanitizeLanguageState(null),
      recent_events: [],
      updated_at: 0,
    };

    // A pending panel reset is authoritative. Otherwise rank language sources
    // and break equal-rank conflicts by the newest state. Candidate streaks
    // are copied from at most one winner and are never added together.
    if (!pendingResetState) {
      for (const candidate of states.slice(1)) {
        state.phase = Math.max(state.phase, candidate.phase);
        const sourceRank = source => ({
          detected: 4, operator_seed: 3, legacy: 2, provisional: 1,
        }[source] || 0);
        if (candidate.language && (
          !state.language
          || sourceRank(candidate.language_source) > sourceRank(state.language_source)
          || (
            sourceRank(candidate.language_source) === sourceRank(state.language_source)
            && candidate.updated_at > state.updated_at
          )
        )) {
          Object.assign(state, sanitizeLanguageState(candidate));
        }
        state.updated_at = Math.max(state.updated_at, candidate.updated_at);
        state.recent_events = [...new Set([
          ...state.recent_events,
          ...candidate.recent_events,
        ])].slice(-this.maxRecentEvents);
      }
    }
    for (const oldKey of existingKeys) {
      if (oldKey !== canonicalKey) delete this.contacts[oldKey];
    }
    for (const [aliasKey, targetKey] of Object.entries(this.aliases)) {
      if (existingKeys.includes(targetKey)) this.aliases[aliasKey] = canonicalKey;
    }
    for (const identityKey of identityKeys) this.aliases[identityKey] = canonicalKey;
    this.contacts[canonicalKey] = state;

    Object.assign(state, reduceLanguageState(state, {
      detectedLanguage,
      languageEvidence,
      provisionalLanguage,
    }));
    const language = state.language || this.defaultLanguage;

    let responseKey;
    if (kind === 'call') {
      responseKey = 'call';
      state.phase = state.phase === 0 ? 1 : 2;
    } else if (state.phase === 0) {
      responseKey = 'step1';
      state.phase = 1;
    } else {
      responseKey = 'step2';
      state.phase = 2;
    }

    state.recent_events.push(eventKey);
    if (state.recent_events.length > this.maxRecentEvents) {
      state.recent_events.splice(0, state.recent_events.length - this.maxRecentEvents);
    }
    state.updated_at = this.now();
    delete state.reset_pending;
    const persisted = this.save();

    return {
      duplicate: false,
      phase: state.phase,
      responseKey,
      language,
      languageSource: state.language_source || null,
      languageCandidate: state.language_candidate || null,
      languageCandidateStreak: state.language_candidate_streak || 0,
      contactKey: canonicalKey,
      persisted,
    };
  }
}

File: interaction_state.py
"""Persistent, privacy-conscious interaction routing for Telegram.

The state machine is deliberately saturated for content: after the first
interaction every text or media interaction receives step 2 forever.  Every
distinct call attempt receives the special call response, regardless of the
current phase, while still advancing the contact's interaction phase.
"""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
import logging
import os
from pathlib import Path
import time
from typing import Callable

from language_adaptation import (
    VALID_LANGUAGES,
    reduce_language_state,
    sanitize_language_state,
)


VALID_KINDS = frozenset({"call", "content"})


@dataclass(frozen=True)
class InteractionDecision:
    duplicate: bool
    phase: int
    response_key: str | None
    language: str
    persisted: bool = True


class PersistentInteractionState:
    """Claims unique inbound events and persists a per-contact phase.

    Raw contact and event identifiers are never written to disk.  Their hashes
    are sufficient for lookup and deduplication while avoiding a readable list
    of customer identifiers in the state file.
    """

    def __init__(
        self,
        file_path: str | Path,
        *,
        default_language: str = "es",
        max_recent_events: int = 256,
        now: Callable[[], float] = time.time,
        logger: logging.Logger | None = None,
    ) -> None:
        if default_language not in VALID_LANGUAGES:
            raise ValueError("default_language must be es, en, or fr")
        if max_recent_events < 1:
            raise ValueError("max_recent_events must be positive")

        self.file_path = Path(file_path)
        self.default_language = default_language
        self.max_recent_events = max_recent_events
        self.now = now
        self.logger = logger or logging.getLogger(__name__)
        self._contacts: dict[str, dict] = {}
        self._load()

    @staticmethod
    def _fingerprint(namespace: str, value: object) -> str:
        material = f"{namespace}\0{value}".encode("utf-8", errors="strict")
        return hashlib.sha256(material).hexdigest()

    def _load(self) -> None:
        if not self.file_path.exists():
            return
        try:
            parsed = json.loads(self.file_path.read_text(encoding="utf-8"))
            if not isinstance(parsed, dict):
                raise ValueError("state root is not an object")
            contacts = parsed.get("contacts", {})
            if not isinstance(contacts, dict):
                raise ValueError("contacts is not an object")

            clean: dict[str, dict] = {}
            for contact_key, raw in contacts.items():
                if not isinstance(contact_key, str) or not isinstance(raw, dict):
                    continue
                phase = raw.get("phase", 0)
                language = raw.get("language")
                events = raw.get("recent_events", [])
                if not isinstance(phase, int) or phase not in (0, 1, 2):
                    continue
                if language not in VALID_LANGUAGES:
                    language = None
                if not isinstance(events, list):
                    events = []
                events = [item for item in events if isinstance(item, str)]
                language_state = sanitize_language_state(raw)
                clean[contact_key] = {
                    "phase": phase,
                    **language_state,
                    "recent_events": events[-self.max_recent_events :],
                    "updated_at": raw.get("updated_at", 0),
                }
            self._contacts = clean
        except (OSError, ValueError, json.JSONDecodeError):
            self.logger.error("Interaction state could not be loaded; starting empty")
            self._contacts = {}

    def _save(self) -> bool:
        try:
            self.file_path.parent.mkdir(parents=True, exist_ok=True)
            temp_path = self.file_path.with_suffix(self.file_path.suffix + ".tmp")
            payload = {"version": 1, "contacts": self._contacts}
            temp_path.write_text(
                json.dumps(payload, ensure_ascii=False, separators=(",", ":")),
                encoding="utf-8",
            )
            os.replace(temp_path, self.file_path)
            return True
        except OSError:
            self.logger.error("Interaction state could not be persisted")
            return False

    def register(
        self,
        *,
        contact_id: object,
        event_id: object,
        kind: str,
        detected_language: str | None = None,
        language_evidence: dict[str, object] | None = None,
        provisional_language: str | None = None,
    ) -> InteractionDecision:
        """Atomically claims an event before any outbound delivery occurs."""

        if contact_id in (None, ""):
            raise ValueError("contact_id is required")
        if event_id in (None, ""):
            raise ValueError("event_id is required")
        if kind not in VALID_KINDS:
            raise ValueError("kind must be call or content")
        if provisional_language not in VALID_LANGUAGES:
            provisional_language = None

        contact_key = self._fingerprint("contact", contact_id)
        event_key = self._fingerprint("event", event_id)
        state = self._contacts.get(contact_key)
        if state is None:
            state = {
                "phase": 0,
                **sanitize_language_state(None),
                "recent_events": [],
                "updated_at": 0,
            }
            self._contacts[contact_key] = state

        recent_events = state["recent_events"]
        if event_key in recent_events:
            return InteractionDecision(
                duplicate=True,
                phase=state["phase"],
                response_key=None,
                language=state.get("language") or self.default_language,
            )

        reduced_language = reduce_language_state(
            state,
            detected_language=detected_language,
            language_evidence=language_evidence,
            provisional_language=provisional_language,
        )
        state.update(reduced_language)
        effective_language = state.get("language") or self.default_language

        previous_phase = state["phase"]
        if kind == "call":
            response_key = "call"
            next_phase = 1 if previous_phase == 0 else 2
        elif previous_phase == 0:
            response_key = "step1"
            next_phase = 1
        else:
            # Every later content interaction uses step 2.
            response_key = "step2"
            next_phase = 2

        recent_events.append(event_key)
        if len(recent_events) > self.max_recent_events:
            del recent_events[: -self.max_recent_events]
        state["phase"] = next_phase
        state["updated_at"] = self.now()
        persisted = self._save()

        return InteractionDecision(
            duplicate=False,
            phase=next_phase,
            response_key=response_key,
            language=effective_language,
            persisted=persisted,
        )

    def preview(
        self,
        *,
        contact_id: object,
        event_id: object,
        kind: str,
        detected_language: str | None = None,
        language_evidence: dict[str, object] | None = None,
        provisional_language: str | None = None,
    ) -> InteractionDecision:
        """Calcula la respuesta sin consumir la interacción todavía.

        El dispatcher la confirma con ``register`` sólo después de completar
        la entrega, de modo que un fallo de Telegram pueda reintentarse.
        """

        if contact_id in (None, ""):
            raise ValueError("contact_id is required")
        if event_id in (None, ""):
            raise ValueError("event_id is required")
        if kind not in VALID_KINDS:
            raise ValueError("kind must be call or content")
        if provisional_language not in VALID_LANGUAGES:
            provisional_language = None

        contact_key = self._fingerprint("contact", contact_id)
        event_key = self._fingerprint("event", event_id)
        state = self._contacts.get(contact_key)
        phase = state["phase"] if state else 0
        language = state.get("language") if state else None
        recent_events = state.get("recent_events", []) if state else []
        if event_key in recent_events:
            return InteractionDecision(
                duplicate=True,
                phase=phase,
                response_key=None,
                language=language or self.default_language,
            )

        reduced_language = reduce_language_state(
            state,
            detected_language=detected_language,
            language_evidence=language_evidence,
            provisional_language=provisional_language,
        )
        effective_language = reduced_language["language"] or self.default_language
        if kind == "call":
            response_key = "call"
        else:
            response_key = "step1" if phase == 0 else "step2"
        return InteractionDecision(
            duplicate=False,
            phase=1 if phase == 0 else 2,
            response_key=response_key,
            language=effective_language,
        )
